src/interface/variable_pack.hpp

```diff
@@ -49,6 +49,53 @@ using SwarmVarList = std::forward_list<std::shared_ptr<ParticleVariable<T>>>;
 // the pairs represent interval (inclusive) of those indices
 using IndexPair = std::pair<int, int>;
 
+// Used for storing the shapes of variable fields
+using Shape = std::vector<int>;
+
+// Index arbitrary rank fields into flattened indices in a VariablePack
+class FlatIdx {
+ public:
+  FlatIdx(std::vector<int> shape, int offset)
+      : shape_("shape", shape.size()), offset_(offset) {
+    for (int i = 0; i < shape.size(); ++i) {
+      shape_(i) = shape[i];
+    }
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  int DimSize(int i) const { return shape_(i); }
+
+  template <typename... Ts>
+  KOKKOS_INLINE_FUNCTION int operator()(Ts... idx_pack) const {
+    std::vector<int> indices = {idx_pack...};
+
+    // Check that the correct dimensionality is being specified
+    if (indices.size() != shape_.size()) {
+      PARTHENON_THROW("Wrong number of indices for variable.")
+    }
+
+    // Find the flat index from the indices assuming fastest moving index is
+    // the rightmost index
+    int idx = 0;
+    if (indices.size() > 0) {
+      for (int idim = indices.size() - 1; idim >= 0; --idim) {
+        if (indices[idim] >= shape_(idim)) {
+          PARTHENON_THROW("Index " + std::to_string(indices[idim]) +
```

Comment:
On GPU this should be `PARTHENON_FAIL`. C++ error patterns don't work on device.

Reply:
I am not that we want to have this error checking in there if it is going to be repeatedly called on the GPU. What do you think?

Reply:
In that case use `PARTHENON_DEBUG_FAIL`... that way it'll be hidden behind preprocessor macros and will be absent in production builds.

Reply:
Switched.